### ISSUE TITLE
Cache invalidation on local dependencies when using docker

### ISSUE THREAD

**Issue by hxyannay** [astral-sh/uv#9191]
Hello.

I am having difficulties using uv in docker. My project structure has two python libraries, one is an API server, and the other is an internal library.
The problem I am encountering is that after building, when running the image, the `uv run` reinstalls my internal libraries, with the following message: `Requirement installed, but not fresh: ...`. This is problematic, because as per docker best practices, it is good to ensure that the install actually happens on the build stage, and not in the actual running phase of my application.
When debugging the issue further, I found out that it happens because the ctime of my `pyproject.toml` of the internal library changes between the docker build and the actual run.

I finally validated that symptom using the following dockerfile:
```
FROM ubuntu:22.04

RUN touch a && stat a > /root/stat-a
RUN sleep 5
```
And running: `docker build --no-cache . -t playground && docker run playground sh -c 'cat /root/stat-a && stat a'`
The output that I received was: 
```
  File: a
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: e3h/227d        Inode: 1336246     Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-11-18 10:04:52.462498005 +0000
Modify: 2024-11-18 10:04:52.462498005 +0000
Change: 2024-11-18 10:04:52.462498005 +0000
 Birth: 2024-11-18 10:04:52.462498005 +0000
---
  File: a
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: e3h/227d        Inode: 225316      Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-11-18 10:04:52.000000000 +0000
Modify: 2024-11-18 10:04:52.000000000 +0000
Change: 2024-11-18 10:04:57.680498008 +0000
 Birth: 2024-11-18 10:04:57.680498008 +0000
```

As seen, two things are changing between the build and run - both the inode number, and the Change & Birth times.
This seems like an odd behavior of docker, but as I dived deeper it seems inherent to the layering system of docker.

I think that the solution to that issue should come from uv's side, as the usage for similar type of projects in docker should be very common.

A possible approach that would solve both the cache freshness problem and would work with docker's odd behavior would be using checksums for local dependencies freshness check.

Some additional information:
- uv version: v0.5.1
- Platform: MacOS with Docker Desktop

Thanks in advance!
Yannay

**Comment by hxyannay** [comment]
Hey,
So I worked a bit to reproduce it, created [this](https://github.com/hxyannay/uv-docker-example) fork of the example from the official documentation of uv
Added [one commit](https://github.com/astral-sh/uv-docker-example/commit/43a86be439084d53d9bc40171d4fb5dee5fdc365) that changes the command to use `uv run` instead of modifying the `ENV` (which is better anyways in my opinion).

When running the build and then run:
```
$ docker build -t uv-example .
$ docker run uv-example
Uninstalled 1 package in 0.68ms
Installed 2 packages in 10ms
...
```
As seen in the last two rows, it uninstalled and reinstalled the package of the example, as the symptoms that I encountered.

Thanks for responding so fast!

**Comment by charliermarsh** [comment]
Thanks! Interestingly, with that example I see:

```
❯ docker run uv-example
Installed 1 package in 6ms
Bytecode compiled 1143 files in 49ms
INFO     Using path src/uv_docker_example
INFO     Resolved absolute path /app/src/uv_docker_example
```

Am I misunderstanding the output, or does it look like a platform difference?

**Comment by hxyannay** [comment]
Yeah it does seem different, I am running this on MacOS with docker desktop.
Though in your output it still shows that it installed a package, is it the expected behavior? I would expect it to not install any package as we run `uv sync` in the dockerfile.

**Comment by hxyannay** [comment]
This is very odd then. 
I am running on M2 with macOS Sequoia 15.1 and Docker Desktop v4.35.1 with docker engine v27.3.1

**Comment by zanieb** [comment]
Are ya'll definitely testing the same uv version? The image could be cached if you're just using that example repository.

**Comment by charliermarsh** [comment]
So in my case at least, the installed package in `Installed 1 package` is Ruff, due to the use of `--no-dev` when syncing the Dockerfile. (`uv run` will sync dev by default.)

**Comment by charliermarsh** [comment]
Have you considered using `uv run --frozen` in your `CMD` instead?

**Comment by charliermarsh** [comment]
(I'm also on Docker version 27.3.1. M3, Sonoma 14.5.)

**Comment by hxyannay** [comment]
Still the same when running with `--frozen` flag.
The uv version from inside the container is `0.5.3`.
Could it be related to the VM options in docker?
I am using Apple Virtualization framework with Rosetta enabled and VirtioFS for file sharing implementation.

**Comment by hxyannay** [comment]
> (I'm also on Docker version 27.3.1. M3, Sonoma 14.5.)

Hey @charliermarsh, is there any more information I can share that will help you reproduce it locally?

**Comment by helmut-hoffer-von-ankershoffen** [comment]
Same issue on my side - following [this example](https://github.com/astral-sh/uv-docker-example/blob/main/Dockerfile) leads to `Uninstalled 1 package [..] Installed 1 package`. Using --verbose on uv run surfaces "Requirement installed, but not fresh", with the requirement being the project itself.

**Comment by czechnology** [comment]
My colleagues using Docker Desktop on Windows are encountering this as well.
Is there a way to bypass or "trick" the freshness check? How does uv decide whether a package is not fresh and requires rebuilding?

**Comment by stisti** [comment]
Me and other my team mates are encountering this issue. Everyone with the issue is running Docker Desktop, either on macOS or Windows. Team mates who have installed Docker community edition in WSL2 do not have this issue. We noticed the issue rather quickly after we switched to using uv because we:
1. run docker build where Dockerfile runs uv to install dependencies using `uv sync --frozen`
2. using the image from step 1, we `docker run --rm --network none uv run --frozen pytest` (and pylint and other tools)

The step 2 fails for people using Docker Desktop because uv tries to rebuild the project and download some libraries (setuptools). Step 2 succeeds for people using Docker community edition.

Of course we can work around the problem by removing `--network none` but it revealed this problem that uv tries to rebuild the python project during docker run.

**Comment by Nugine** [comment]
Is there any workaround? I don't want to re-install anything on container startup.

**Comment by jmesterh** [comment]
We are switching to Chainguard images, which requires a multi-stage build where you build in one stage with the "-dev" container, and copy the build artifacts into the final stage which doesn't have /bin/sh, etc. 

The Docker COPY command doesn't preserve timestamps, so in the final stage when it tries to run the program:

```
CMD [ "uv", "run", ... ]
```

uv says the "Requirement installed, but not fresh" for the project itself and it tries to reinstall. 

We can't just call the project directly vs `uv run` because the library we use (Prefect) calls `uv run` internally with lots of arguments l don't have control over like `--python` and `--with`

I would like an environment variable that tells uv not to try and do *anything* and just run the program. 

*edit* it seems this combination of environment variables at the end of the Dockerfile works:
```
ENV UV_NO_MANAGED_PYTHON=true \
    UV_SYSTEM_PYTHON=true \
    UV_LOCKED=1 \
    UV_NO_SYNC=1 \
    UV_NO_CACHE=1 \
    UV_NO_DEV=1
```
